format/FlightSql.proto

```diff
@@ -0,0 +1,440 @@
+ *  update_rule: uint1,
+ *  delete_rule: uint1
+ * >
+ * The returned data should be ordered by pk_catalog_name, pk_schema_name, pk_table_name, pk_key_name, then key_sequence.
+ * update_rule and delete_rule returns a byte that is equivalent to actions:
+ *    - 0 = CASCADE
+ *    - 1 = RESTRICT
+ *    - 2 = SET NULL
+ *    - 3 = NO ACTION
+ *    - 4 = SET DEFAULT
+ */
+message CommandGetImportedKeys {
+  option (experimental) = true;
+
+  // Specifies the catalog to search for the primary key table.
+  google.protobuf.StringValue catalog = 1;
+
+  // Specifies the schema to search for the primary key table.
+  google.protobuf.StringValue schema = 2;
+
+  // Specifies the primary key table to get the foreign keys for.
+  string table = 3;
+}
+
+// SQL Execution Action Messages
+
+/*
+ * Request message for the "CreatePreparedStatement" action on a Flight SQL enabled backend.
+ */
+message ActionCreatePreparedStatementRequest {
+  option (experimental) = true;
+
+  // The valid SQL string to create a prepared statement for.
+  string query = 1;
+}
+
+/*
+ * Wrap the result of a "GetPreparedStatement" action.
+ */
+message ActionCreatePreparedStatementResult {
+  option (experimental) = true;
+
+  // Opaque handle for the prepared statement on the server.
+  bytes prepared_statement_handle = 1;
+
+  // If a result set generating query was provided, dataset_schema contains the 
+  // schema of the dataset as described in Schema.fbs::Schema, it is serialized as an IPC message.
+  bytes dataset_schema = 2;
+
+  // If the query provided contained parameters, parameter_schema contains the 
+  // schema of the expected parameters as described in Schema.fbs::Schema, it is serialized as an IPC message.
+  bytes parameter_schema = 3;
+}
+
+/*
+ * Request message for the "ClosePreparedStatement" action on a Flight SQL enabled backend.
+ * Closes server resources associated with the prepared statement handle.
+ */
+message ActionClosePreparedStatementRequest {
+  option (experimental) = true;
+
+  // Opaque handle for the prepared statement on the server.
+  bytes prepared_statement_handle = 1;
+}
+
+
+// SQL Execution Messages.
+
+/*
+ * Represents a SQL query. Used in the command member of FlightDescriptor
+ * for the following RPC calls:
+ *  - GetSchema: return the schema of the query.
+ *  - GetFlightInfo: execute the query.
+ */
+message CommandStatementQuery {
+  option (experimental) = true;
+
+  // The SQL syntax.
+  string query = 1;
+}
+
+/*
+ * Represents an instance of executing a prepared statement. Used in the command member of FlightDescriptor for
+ * the following RPC calls:
+ *  - DoPut: bind parameter values. All of the bound parameter sets will be executed as a single atomic execution.
+ *  - GetFlightInfo: execute the prepared statement instance.
+ */
+message CommandPreparedStatementQuery {
+  option (experimental) = true;
+
+  // Opaque handle for the prepared statement on the server.
+  bytes prepared_statement_handle = 1;
+}
```

Comment:
Was the client handle not still necessary here and below? It's still used in the client code.

Reply:
It was actually needed for CommandStatementQuery and not for CommandPreparedStatementQuery nor CommandPreparedStatementUpdate.
I mistakenly removed it from CommandStatementQuery in the last commit and now fixed

Reply:
Thanks. I suppose I don't see what the intended use or semantics of the field are. It's used to index into a cache, but that cache seems like it should be indexed by query, not by client handle (else if a client maliciously or accidentally reuses a handle they can get the wrong FlightInfo). And if it's unique, then why cache anything anyways? It'll never come up again.

Reply:
Hi @lidavidm ,

Just to make sure, we are actually talking about `CommandStatementQuery`, right?
If so, we could not index the cache by using the query itself, as it can have the same query running concurrently.

What was confusing about this is that we are using CommandStatementQuery as both the command in getFlightInfo (which needs the query) and ticket when calling getStream (which needs the handle).

I think a solution to that is to split CommandStatementQuery into two messages:
```proto
message CommandStatementQuery {
  option (experimental) = true;

  // The SQL syntax.
  string query = 1;
}

message TicketStatementQuery {
  option (experimental) = true;
  
  // Unique identifier for the instance of the statement to execute.
  bytes statement_handle = 1;
}
```

`CommandStatementQuery` would be used by the client when calling getFlightInfo, the server would create and cache the statement and return a FlightInfo with a ticket containing `TicketStatementQuery` .

Reply:
Ah - I see now, thanks. Yes, this is the intent of the Ticket, and I agree splitting messages like this makes more sense. Also, I think it should be stated (at least for now) that clients should not attempt to parse TicketStatementQuery / clients should treat the Ticket payload as opaque.

Reply:
@lidavidm , we just made this change! Can you take a look please? :)

Reply:
Thanks! I've kicked off CI, I will take a look later this evening.